src/main.rs

```diff
@@ -61,32 +56,32 @@ async fn main() -> Result<()> {
         "0x0546a164c8d10fd38652b6426ef7be159965deb9a0cbf3e8a899f8a42fd86761",
     )
     .unwrap();
-
-    let contract_caller_a = ContractA::new_caller(contract_address, provider).await?;
-
-    // TODO: if we can handle the SingleOwnerAccount lifetime, it will be easier to
-    // only pass the account. And not account_address + signer.
-    let contract_invoker =
-        ContractA::new_invoker(contract_address, provider2, account_address, signer).await?;
-
-    let contract_caller_b = ContractB::new_caller(contract_address, provider3).await?;
-
-    contract_invoker
+    let chain_id = provider.chain_id().await?;
+    let account = SingleOwnerAccount::new(&provider, signer, account_address, chain_id);
+    let contract_a = ContractA::new(contract_address, &provider)
+        .with_account(&account)
+        .await?;
+
+    // 1) you can either call using explicit type ( without `with_account` )
+    // TODO: Is there any way that don't need explicit type and still keep A generic even without account?
```

Comment:
This is exactly what I want to avoid. :+1: Totally agree with you, if someone want to make a simple call, we don't have to bother him giving all the details of the implementation it will not event use.

Reply:
As the compiler can't infer the type, I'm not sure it's possible to do so. Or at least not in that configuration.